src/Prefix.jl

```diff
@@ -271,62 +271,65 @@ the environment variables that will be defined within the sandbox environment.
-    destdir = joinpath(build_path, nonce, "destdir")
-    metadir = joinpath(build_path, nonce, "metadir")
-    wrapperdir = joinpath(build_path, nonce, "compiler_wrappers")
+    srcdir = joinpath(workspace, "srcdir")
+    destdir = joinpath(workspace, "destdir")
+    metadir = joinpath(workspace, "metadir")
+    wrapperdir = joinpath(workspace, "compiler_wrappers")
     mkdir.((srcdir, destdir, metadir))
 
     # For each source path, unpack it
-    for (src_path, src_hash) in zip(src_paths, src_hashes)
-        if isdir(src_path)
-            # Chop off the `.git` at the end of the src_path
-            repo_dir = joinpath(srcdir, basename(src_path)[1:end-4])
-            LibGit2.with(LibGit2.clone(src_path, repo_dir)) do repo
-                LibGit2.checkout!(repo, src_hash)
+    for source in sources
+        targetdir = joinpath(srcdir, source.unpack_target)
+        mkpath(targetdir)
+        if isdir(source.path)
+            # Chop off the `.git` at the end of the source.path
+            repo_dir = joinpath(targetdir, basename(source.path)[1:end-4])
+            LibGit2.with(LibGit2.clone(source.path, repo_dir)) do repo
+                LibGit2.checkout!(repo, source.hash)
             end
         else
             # Extract with host tools because it is _much_ faster on e.g. OSX.
             # If this becomes a compatibility problem, we'll just have to install
             # our own `tar` and `unzip` through BP as dependencies for BB.
             tar_extensions = [".tar", ".tar.gz", ".tgz", ".tar.bz", ".tar.bz2",
                               ".tar.xz", ".tar.Z", ".txz"]
-            cd(srcdir) do
-                if any(endswith(src_path, ext) for ext in tar_extensions)
+            cd(targetdir) do
+                if any(endswith(source.path, ext) for ext in tar_extensions)
                     if verbose
-                        @info "Extracting tarball $(basename(src_path))..."
+                        @info "Extracting tarball $(basename(source.path))..."
                     end
                     tar_flags = verbose ? "xvof" : "xof"
-                    run(`tar $(tar_flags) $(src_path)`)
-                elseif endswith(src_path, ".zip")
+                    run(`tar $(tar_flags) $(source.path)`)
+                elseif endswith(source.path, ".zip")
                     if verbose
-                        @info "Extracting zipball $(basename(src_path))..."
+                        @info "Extracting zipball $(basename(source.path))..."
                     end
-                    run(`unzip -q $(src_path)`)
+                    run(`unzip -q $(source.path)`)
                else
                    if verbose
-                       @info "Copying in $(basename(src_path))..."
+                       @info "Copying in $(basename(source.path))..."
                    end
-                   cp(src_path, basename(src_path))
+                   cp(source.path, joinpath(pwd(), source.unpack_target, basename(source.path)))
```

Comment:
So if we have an un-extractable file (like a `.dmg` as a source, but we provide an `unpack_target`, it gets put into `${SRCDIR}/${unpack_target}/${basename}`?

Reply:
Well, I may have swapped the two arguments :sweat_smile:

Reply:
Actually, there was no need of the whole `joinpath` thing here, I only had to change `src_path` to `source.path`.  There is a `cd()` above that moves the current directory to the desired place directory.